File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        return find(bloomDay,m,k);
    }
    private static int canMake(int[] bloomDay,int mid,int k){
        int bouqCount=0;
        int c=0;
        for(int i=0;i<bloomDay.length;i++){
            if(bloomDay[i]<=mid){
                c++;
            }else{
                c=0;
            }
            if(c==k){
                bouqCount++;
                c=0;
            }
        }

        return bouqCount;
    }

    private static int find(int[] bloomDay,int m,int k){
        int start=1;
        int end=-1;
        for(int day:bloomDay){
            if(day>=end){
                end=day;
            }
        }
        int ans=-1;
        while(start<=end){
            int mid=start+(end-start)/2;
            
            if(canMake(bloomDay,mid,k)>=m){
                ans=mid;
                end=mid-1;
            }else{
                start=mid+1;
            }
        }
        return ans;
    }
}

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool canBloom(vector<int>& bloomDays, int m, int k, int day) {
        int consecutive = 0;
        int bouquets = 0;

        for (int i = 0; i < bloomDays.size(); i++) {
            if (bloomDays[i] <= day) {
                consecutive++;
            } else {
                consecutive = 0;
            }

            if (consecutive == k) {
                bouquets++;
                consecutive = 0;

                if (bouquets == m)
                    return true;
            }
        }

        return false;
    }

    int minDays(vector<int>& bloomDay, int m, int k) {
        if ((long long)m * k > bloomDay.size())
            return -1;

        int low = INT_MAX;
        int high = INT_MIN;

        for (int d : bloomDay) {
            low = min(low, d);
            high = max(high, d);
        }

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canBloom(bloomDay, m, k, mid)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }
};